spi/control-plane/control-plane-spi/src/main/java/org/eclipse/edc/connector/spi/catalog/DatasetResolver.java

```diff
@@ -0,0 +1,40 @@
+/*
+ *  Copyright (c) 2023 Bayerische Motoren Werke Aktiengesellschaft (BMW AG)
+ *
+ *  This program and the accompanying materials are made available under the
+ *  terms of the Apache License, Version 2.0 which is available at
+ *  https://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  SPDX-License-Identifier: Apache-2.0
+ *
+ *  Contributors:
+ *       Bayerische Motoren Werke Aktiengesellschaft (BMW AG) - initial API and implementation
+ *
+ */
+
+package org.eclipse.edc.connector.spi.catalog;
+
+import org.eclipse.edc.catalog.spi.DataService;
+import org.eclipse.edc.catalog.spi.Dataset;
+import org.eclipse.edc.spi.agent.ParticipantAgent;
+import org.eclipse.edc.spi.query.QuerySpec;
+import org.jetbrains.annotations.NotNull;
+
+import java.util.stream.Stream;
+
+/**
+ * Resolves {@link Dataset}s for the {@link CatalogProtocolService}
+ */
+public interface DatasetResolver {
+
+    /**
+     * Resolves {@link Dataset}s given the {@link ParticipantAgent}, a {@link QuerySpec} and a {@link DataService}
+     *
+     * @param agent the participant agent that requested the dataset.
+     * @param querySpec the query spec for filtering and pagination.
+     * @param dataService the data service to be associated to the datasets.
+     * @return a stream of datasets.
+     */
+    @NotNull
+    Stream<Dataset> query(ParticipantAgent agent, QuerySpec querySpec, DataService dataService);
```

Comment:
Passing in the DataService seems a bit strange. Shouldn't the resolver delegate to another service internally to resolve the DataService? For example, different Datasets may be provided by different DataServices.

Reply:
The `DataService` objects are needed at the `Catalog` level, and for every `Distribution` we'd need the `DataService` id. To avoid this there could be a `DataServiceResolver` service that could be injected both in the `CatalogService` and in the `DatasetResolver` and retrieve all the `DataService` (that, for a connector will be the connector itself).

I would prefer to address this to a separate issue/PR to provide the functionality incrementally.